components/snap_recovery/src/leader_keeper.rs

```diff
@@ -0,0 +1,260 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    cell::RefCell,
+    collections::HashSet,
+    marker::PhantomData,
+    time::{Duration, Instant},
+};
+
+use engine_traits::KvEngine;
+use futures::compat::Future01CompatExt;
+use itertools::Itertools;
+use raftstore::{
+    errors::{Error, Result},
+    store::{Callback, CasualMessage, CasualRouter, SignificantMsg, SignificantRouter},
+};
+use tikv_util::{future::paired_future_callback, timer::GLOBAL_TIMER_HANDLE};
+
+pub struct LeaderKeeper<EK, Router> {
+    router: Router,
+    not_leader: HashSet<u64>,
+
+    _ek: PhantomData<EK>,
+}
+
+#[derive(Default)]
+pub struct StepResult {
+    pub failed_leader: Vec<(u64, Error)>,
+    pub campaign_failed: Vec<(u64, Error)>,
+}
+
+fn ellipse<T: std::fmt::Debug>(ts: &[T], max_len: usize) -> String {
+    if ts.len() < max_len {
+        return format!("{:?}", &ts);
+    }
+    format!("{:?} (and {} more)", &ts[..max_len], ts.len() - max_len)
+}
+
+impl std::fmt::Debug for StepResult {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("StepResult")
+            .field(
+                "failed_leader",
+                &format_args!("{}", ellipse(&self.failed_leader, 8)),
+            )
+            .field(
+                "campaign_failed",
+                &format_args!("{}", ellipse(&self.campaign_failed, 8)),
+            )
+            .finish()
+    }
+}
+
+impl<EK, Router> LeaderKeeper<EK, Router>
+where
+    EK: KvEngine,
+    Router: CasualRouter<EK> + SignificantRouter<EK> + 'static,
+{
+    pub fn new(router: Router, to_keep: impl IntoIterator<Item = u64>) -> Self {
+        Self {
+            router,
+
+            not_leader: to_keep.into_iter().collect(),
+            _ek: PhantomData,
+        }
+    }
+
+    pub async fn wait_until_all_leader_ready(&mut self) {
+        loop {
+            let now = Instant::now();
+            GLOBAL_TIMER_HANDLE
+                .delay(now + Duration::from_secs(30))
+                .compat()
+                .await
+                .expect("wrong with global timer, cannot stepping.");
+            let res = self.step().await;
+            info!("finished leader keeper stepping."; "result" => ?res, "take" => ?(now.elapsed() - Duration::from_secs(30)));
+            if res.failed_leader.is_empty() {
+                return;
+            }
+        }
+    }
+
+    pub async fn step(&mut self) -> StepResult {
+        const CONCURRENCY: usize = 256;
+        let r = RefCell::new(StepResult::default());
+        let success = RefCell::new(HashSet::new());
+        for batch in &self.not_leader.iter().chunks(CONCURRENCY) {
+            let tasks = batch.map(|region_id| async {
+                match self.check_leader(*region_id).await {
+                    Ok(_) => {
+                        success.borrow_mut().insert(*region_id);
```

Comment:
If it's running in a multi-thread pool, this may panic? Maybe a rwlock is better.

Reply:
Generally `RefCell` and the `RaftStoreRouter` will forbid we from running it at a multi-threaded pool. Anyway using Mutex won't make things worse.

Reply:
I see. Now I think it's OK.